Add command for composite commands in devfile.yaml

msivasubramaniaan opened the pull request:
### What does this PR do?
This PR shows the composite commands in the task

### What issues does this PR fix?
https://github.com/eclipse-che/che/issues/21859

### How to test this PR?

- Create a workspace which has `devfile.yaml` with composite commands Either (parallel/sequence) order
- Open Terminal -> Run Task -> devfile
- The composite command item will be shown in the dropdown
 
The below 3 cases needs to be tested

1. parallel - Composite commands should be run parallelly
```
- id: build-and-run
    composite:
      commands:
        - npm-build
        - npm-run
      parallel: true
      label: Build and Run che-code
      group:
        kind: run
        isDefault: true
```
2. sequence - Composite commands should be run one after another
```
- id: build-and-run
    composite:
      commands:
        - npm-build
        - npm-run
      parallel: false
      label: Build and Run che-code
      group:
        kind: run
        isDefault: true
```
3. if no reference of the commands - For example composite commands are `npm-install` and `npm-build`. But there were no reference for `npm-install` then that composite commands are not shown in the dropdown

### Does this PR contain changes that override default upstream Code-OSS behavior?
- [ ] the PR contains changes in the [code](https://github.com/che-incubator/che-code/tree/main/code) folder (you can skip it if your changes are placed in a che extension )
- [ ] the corresponding items were added to the [CHANGELOG.md](https://github.com/che-incubator/che-code/blob/main/.rebase/CHANGELOG.md) file
- [ ] rules for automatic `git rebase` were added to the [.rebase](https://github.com/che-incubator/che-code/tree/main/.rebase) folder

msivasubramaniaan committed:
fix: added functionality for composite command support and code formatting

msivasubramaniaan committed:
devfile updated for testing purpose

msivasubramaniaan committed:
updated the devfile.yaml for added composite command

msivasubramaniaan committed:
updated the taskProvider.ts

msivasubramaniaan committed:
updated the taskprovider

msivasubramaniaan committed:
added nested composite command supoort and removed npm-install command for testing purpose

msivasubramaniaan committed:
formatted the code and revert the devfile.yaml

RomanNikitenko commented:
@msivasubramaniaan 
thank you for the adding this functionality!

please provide info to the `How to test this PR` section
thanks in advance!

RomanNikitenko commented:
@msivasubramaniaan 
thank you for the updated `How to test this PR` section!

I would add different use cases that should be tested for this functionality, like:
- sample with commands that should be run `in sequence`
- sample with commands that should be run `in parallel`
- any other use cases that you tested when you worked on this functionality

`How to test this PR` section is useful when a PR is under review.
This section is also helpful when some changes are added in a month/year - to test if all use cases still work as expected and there is no a regression.

RomanNikitenko commented on devfile.yaml:
```diff
@@ -128,4 +128,4 @@ commands:
         git stash save --include-untracked
         git fetch upstream-code main
-        ./rebase.sh
+        ./rebase.sh
```
please revert this change

msivasubramaniaan committed:
revert the devfile.yaml

RomanNikitenko commented:
@msivasubramaniaan 
the image for current PR was not built and is absent at the moment - so - it's not possible to test your changes...
I've restarted the job - waiting for the image...

RomanNikitenko commented:
@msivasubramaniaan 
there are still no published images for the current PR 

I've fixed the problem:  https://github.com/eclipse-che/che/issues/23663
Please rebase your branch to get published images for your PR

RomanNikitenko commented on .github/workflows/pull-request-check-publish.yml:
```diff
@@ -21,6 +21,18 @@ jobs:
+        with:
+          registry: quay.io
+          username: ${{ secrets.QUAY_PULL_REQUESTS_USERNAME }}
+          password: ${{ secrets.QUAY_PULL_REQUESTS_PASSWORD }}
```
it's strange that these changes are displayed as changes going to be merged to the `main`...
something wrong with rebasing?

msivasubramaniaan replied:
no I just rebase it. Don't know how it displayed as a change

RomanNikitenko replied:
<img width="561" height="597" alt="image" src="https://github.com/user-attachments/assets/330a4bf0-ae90-4857-bc8d-98c77bb0c8be" />

RomanNikitenko replied:
my commits are displayed in your PR

msivasubramaniaan committed:
Merge branch 'main' into add-command-for-composite-commands-in-devfile

RomanNikitenko reviewed (commented):
I've tested the changes.
In general it works, at the same time I've detected some problems.

### Output is lost
For example, I created simple commands that just check `npm` and `node` versions.
It works well when the commands are executed in sequence.
But an output for the first command is lost when the commands are executed in parallel:

#### Steps to reproduce:
1. Start workspace - [click here](https://workspaces.openshift.com#https://github.com/RomanNikitenko/web-nodejs-sample/tree/composite-command?editor-image=quay.io/che-incubator-pull-requests/che-code:pr-601-amd64)
2. Terminal => Run Task => devfile => `Sequential: Check versions`
3. Terminal => Run Task => devfile => `Parallel: Check versions`

<img width="498" height="188" alt="Screenshot 2025-12-02 at 12 17 25" src="https://github.com/user-attachments/assets/b30c4a63-1ba8-4b05-b47d-1a32a0d1bc20" />

### Variables are not resolved
Another problem that I've faced - variables, like `${PROJECT_SOURCE}` are not resolved for the composite commands.

#### Steps to reproduce:
1. Start workspace - [click here](https://workspaces.openshift.com#https://github.com/RomanNikitenko/web-nodejs-sample/tree/composite-command?editor-image=quay.io/che-incubator-pull-requests/che-code:pr-601-amd64)
2. Terminal => Run Task => devfile => `Check versions => Install deps` 
<img width="498" height="139" alt="Screenshot 2025-12-02 at 12 35 08" src="https://github.com/user-attachments/assets/743a00c0-8f62-4dc2-8323-fcb4173f18be" />

msivasubramaniaan committed:
support parallel wait and PROJECT_SOURCE and updated devfile.yaml for testing purpose

msivasubramaniaan committed:
updated devfile.yaml

msivasubramaniaan committed:
update devfile.yaml for PROJECT_SOURCE

msivasubramaniaan committed:
revert devfile.yaml

msivasubramaniaan commented:
> I've tested the changes. In general it works, at the same time I've detected some problems.
> 
> ### Output is lost
> For example, I created simple commands that just check `npm` and `node` versions. It works well when the commands are executed in sequence. But an output for the first command is lost when the commands are executed in parallel:
> 
> #### Steps to reproduce:
> 1. Start workspace - [click here](https://workspaces.openshift.com#https://github.com/RomanNikitenko/web-nodejs-sample/tree/composite-command?editor-image=quay.io/che-incubator-pull-requests/che-code:pr-601-amd64)
> 2. Terminal => Run Task => devfile => `Sequential: Check versions`
> 3. Terminal => Run Task => devfile => `Parallel: Check versions`
> 
> <img alt="Screenshot 2025-12-02 at 12 17 25" width="498" height="188" src="https://private-user-images.githubusercontent.com/5676062/521269181-b30c4a63-1ba8-4b05-b47d-1a32a0d1bc20.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjQ2ODUxNjcsIm5iZiI6MTc2NDY4NDg2NywicGF0aCI6Ii81Njc2MDYyLzUyMTI2OTE4MS1iMzBjNGE2My0xYmE4LTRiMDUtYjQ3ZC0xYTMyYTBkMWJjMjAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MTIwMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTEyMDJUMTQxNDI3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjRhYjI0NTBkNWIwMGJjNmViNjAzYmU1MGQyZmM2OWQxZjcxNWFhYjYzNjc2YzdhY2FkYjgwNGY2NjBlZTRiMyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.iKZfJulP_HRxL7WgS9QqMBkxNyo9XCgUqJLPumxpAXA">
> ### Variables are not resolved
> Another problem that I've faced - variables, like `${PROJECT_SOURCE}` are not resolved for the composite commands.
> 
> #### Steps to reproduce:
> 1. Start workspace - [click here](https://workspaces.openshift.com#https://github.com/RomanNikitenko/web-nodejs-sample/tree/composite-command?editor-image=quay.io/che-incubator-pull-requests/che-code:pr-601-amd64)
> 2. Terminal => Run Task => devfile => `Check versions => Install deps`
> 
> <img alt="Screenshot 2025-12-02 at 12 35 08" width="498" height="139" src="https://private-user-images.githubusercontent.com/5676062/521273223-743a00c0-8f62-4dc2-8323-fcb4173f18be.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjQ2ODUxNjcsIm5iZiI6MTc2NDY4NDg2NywicGF0aCI6Ii81Njc2MDYyLzUyMTI3MzIyMy03NDNhMDBjMC04ZjYyLTRkYzItODMyMy1mY2I0MTczZjE4YmUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MTIwMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTEyMDJUMTQxNDI3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OTFhYWVmZmQ4NmFkNWViN2VhYTNmZDA4Y2MyZTFiYzA3NmViYTA1NTg2YzBhYTVhMTZiNDgzNGM5NTkxMThjYyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.vSzGSdBDXRm7FyfFTBDJ_6cpuQd8RntxXlai8GKfABo">

@RomanNikitenko Thanks for the testing. I've fixed the mentioned issue and updated the code. Please have a look 
<img width="821" height="302" alt="image" src="https://github.com/user-attachments/assets/e1eb1421-9e40-4321-a170-541995926b6a" />

RomanNikitenko reviewed (approved):
I tested the functionality using samples I prepared in https://github.com/che-incubator/che-code/pull/601#pullrequestreview-3529416213

I didn't detect any problem.
Thanks @msivasubramaniaan !

RomanNikitenko commented:
@msivasubramaniaan 
The license check is failing for your PR, 
but no dependencies were added/changed within the current PR.

So, I believe - it's ready to merge!
